ci/src/main.rs

```diff
@@ -0,0 +1,213 @@
+//! Continuous Integration: rust-bitcoin test runner.
+//!
+//! This program is used by our CI pipeline to test the library. It can, of course, also be used
+//! from the command line also.
+//!
+//! # Hints and tips
+//!
+//! - Build and run (from the repo root directory)
+//!
+//!    `cd ci; cargo build --target-dir target; cd ..; ci/target/debug/ci [OPTIONS]`
+//!
+//! - If you would like more verbose output from cargo consider setting `CARGO_TERM_VERBOSE`.
+//!
+//! - To debug or just see that the tests you think are running are running consider sending
+//!   stdout /dev/null and use `--debug` to see descriptive output.
+//!
+//! - To control the toolchain use `RUSTUP_TOOLCHAIN=nightly`.
+
+use std::process;
+
+use cmd_lib::*;
+use clap::Parser;
+
+/// The list of individual features that we test.
+const FEATURES: &str = "base64 bitcoinconsensus serde rand secp-recovery";
+
+/// Test runner for the rust-bitcoin library.
+#[derive(Parser, Debug)]
+#[clap(version, about, long_about = None)]
+struct Args {
+    /// Build and test the "no-std" feature.
+    #[clap(long)]
+    no_std: bool,
+
+    /// Control which toolchain is used.
+    #[clap(long, value_parser, default_value = "")]
+    toolchain: String,
+
+    /// Build and test each feature and various combinations of features.
+    #[clap(long)]
+    feature_matrix: bool,
+
+    /// Build the docs.
+    #[clap(long)]
+    docs: bool,
+
+    /// Run the fuzz tests.
+    #[clap(long)]
+    fuzz: bool,
+
+    /// Run the bench marks.
+    #[clap(long)]
+    bench: bool,
+
+    /// Test bitcoin can be used as a dependency.
+    #[clap(long)]
+    as_dependency: bool,
+
+    /// Turn on debug output.
+    #[clap(short, long)]
+    debug: bool,
+}
+
+fn main() -> CmdResult {
+    let args = Args::parse();
+
+    // Debugging output: Use a closure so we do not have to pass `args.debug` around.
+    let debug = |msg| if args.debug {
```

Comment:
Nice trick! Took me a bit to understand but can't figure out a more readable solution that's not too annoying to write.
Maybe just longer doc:

> This closure prints its argument to stderr if debugging is enabled.

Reply:
Yeah this closure is the main thing I'm not happy with and where Bash wins out IMO. What I really wanted was a macro controlled by a flag that when debugging just prints the commands and doesn't run them, that way one can see that the 'script' is running the expected commands. I've found bugs before when the current  script was run from CI with some combination of env vars that led to parts of the code not running.

Reply:
Maybe use the `log` crate? Or if you want something smaller:

```rust
use once_cell::sync::OnceCell;
static DEBUG: OnceCell<bool> = OnceCell::new();

macro_rules! debug {
    ($fstr:lit $(, $arg:tt)*) => {
        if DEBUG.get().expect("debug! called before DEBUG was initialized") {
            eprintln!($fstr $(, $arg)*);
        }
    }
}


fn main() {
    // ...
    let args = Args::parse();
    DEBUG.set(args.debug).expect("DEBUG initialized twice");
    // use `debug!()` instead of `debug()` from here
}
```

Reply:
Ha, so long since I wrote application code that I forgot about logging :)